### PR TITLE
cpu attn mask

### PR DESCRIPTION
**Pull request by ncfrey**
## Description
This pull request introduces significant improvements to the embedding computation in the `UME` model and adds comprehensive tests to ensure consistency across different configurations. The changes primarily address the handling of padding tokens during mean pooling and introduce new test cases to validate the consistency of embeddings generated by the model across devices and configurations.

### Improvements to embedding computation:

* [`src/lobster/model/_ume.py`](diffhunk://#diff-d191d7b2a2e81ee87c07961e919192007358d2a98a5e396cfabba050db44f479L443-R470): Updated the embedding computation logic to use an attention mask for excluding padding tokens from mean pooling. This ensures accurate aggregation by masking out padding tokens, preventing them from affecting the computed embeddings. Additionally, safeguards were added to handle mismatched sequence lengths between the attention mask and embeddings.

### Enhanced test coverage:

* [`tests/lobster/model/test__ume.py`](diffhunk://#diff-3f96a9813768980bff629197578aa724adfac50736c7589ea5e18c7dcc9874adR295-R414): Added a new test method `test_flash_attention_consistency_across_devices` to validate the consistency of embeddings generated with flash attention (GPU) and non-flash attention (CPU). The test includes various modalities (e.g., amino acid, nucleotide, SMILES) and checks cosine similarity to ensure embeddings are nearly identical (>99.9% similarity). It also tests consistency between padded and unpadded architectures.

## Type of Change
- [X] Bug fix
- [ ] New feature  
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring

## Testing
- [X] Tests pass locally
- [X] Added new tests for new functionality
- [X] Updated existing tests if needed

## Checklist
- [X] Code follows style guidelines
- [X] Self-review completed
- [X] Documentation updated if needed
- [X] No breaking changes (or clearly documented)

### CODE REVIEW

File: tests/lobster/model/test__ume.py [hunk 1]
```diff
@@ -292,6 +292,126 @@ def test_embed_sequences_gpu_flash_attn(self):
...
+                    embeddings_gpu = ume_gpu.embed_sequences(sequences, modality, aggregate=True)
+                    embeddings_cpu = ume_cpu.embed_sequences(sequences, modality, aggregate=True)
+
+                # Move to same device for comparison
+                embeddings_gpu = embeddings_gpu.cpu()
+
+                # Check similarity
+                cosine_sim = torch.nn.functional.cosine_similarity(embeddings_gpu, embeddings_cpu, dim=1)
+
+                print(f"  Cosine similarities: {cosine_sim.tolist()}")
```
Comment (karinazad) [inline tests/lobster/model/test__ume.py:357]: might want to remove print statements from tests before merging?

File: src/lobster/model/_ume.py [hunk 2]
```diff
@@ -440,8 +440,34 @@ def embed(
...
         if aggregate:
-            # Use mean pooling over sequence length dimension
-            embeddings = embeddings.mean(dim=1)
+            # Use attention mask to exclude padding tokens from mean pooling
+            attention_mask = x["attention_mask"]
+            if attention_mask.dim() == 3:
+                attention_mask = attention_mask.squeeze(1)  # Remove middle dimension: (batch, seq_len)
+
+            # Ensure attention mask matches embedding sequence length
+            if attention_mask.shape[1] != embeddings.shape[1]:
```
Comment (karinazad) [inline src/lobster/model/_ume.py:449]: how does it happen that the attention mask's shape is not the same as embeddings?

File: tests/lobster/model/test__ume.py (hunk 1 above)
Comment (ncfrey) [reply tests/lobster/model/test__ume.py:357]: i'll replace with logging statements

File: src/lobster/model/_ume.py (hunk 2 above)
Comment (ncfrey) [reply src/lobster/model/_ume.py]: the attention mask is from the unpadded inputs for flash-attn, but the sdpa embeddings are padded so the attention mask has to be updated to match

**Comment by karinazad** [comment]
good fix! we can merge this for now but I'm a bit worried that we're just patching the issue in UME. maybe this should go to modern bert?

File: src/lobster/model/_ume.py [hunk 3]
```diff
@@ -439,9 +439,39 @@ def embed(
         if self.model.config.padding == "unpadded":
             embeddings = embeddings.view(batch_size, seq_len, -1)
 
+        # Handle attention mask for both aggregated and token-level outputs
+        attention_mask = x["attention_mask"]
+        if attention_mask.dim() == 3:
+            attention_mask = attention_mask.squeeze(1)  # Remove middle dimension: (batch, seq_len)
+
+        # Ensure attention mask matches embedding sequence length
```
Comment (karinazad) [inline src/lobster/model/_ume.py:447]: claude suggested:


```
batch_size, mask_seq_len = attention_mask.shape
embed_seq_len = embeddings.shape[1]

if mask_seq_len != embed_seq_len:
    if mask_seq_len > embed_seq_len:
        attention_mask = attention_mask[:, :embed_seq_len]
    else:
        pad_length = embed_seq_len - mask_seq_len
        padding = torch.zeros(batch_size, pad_length, 
                            dtype=attention_mask.dtype, 
                            device=attention_mask.device)
        attention_mask = torch.cat([attention_mask, padding], dim=1)
```

Comment (karinazad) [reply src/lobster/model/_ume.py:447]: nit: could be moved to a helper func something like `_align_attention_mask`

**Review by karinazad — APPROVED** [review]

File: src/lobster/model/_ume.py [hunk 4]
```diff
@@ -439,9 +439,39 @@ def embed(
...
         if aggregate:
-            # Use mean pooling over sequence length dimension
-            embeddings = embeddings.mean(dim=1)
+            # Apply mask and compute mean only over actual tokens
+            masked_embeddings = embeddings * mask
+            sum_embeddings = masked_embeddings.sum(dim=1)  # (batch, hidden_dim)
+            token_counts = mask.sum(dim=1)  # (batch, 1)
+
+            # Avoid division by zero for empty sequences
+            token_counts = torch.clamp(token_counts, min=1.0)
```
[the remark this answers is not in the document]
Comment (ncfrey) [reply src/lobster/model/_ume.py:469]: just safety
